specifyweb/workbench/upload/treerecord.py

```diff
@@ -19,33 +19,138 @@
+        }, "Tree is required"
+        return TreeRankRecord(self.rank_name, self.treedef_id)
+
+def get_tree_model(tree: str):
+    return getattr(models, tree.lower().title() + 'treedefitem')
+
+class TreeRankRecord(NamedTuple):
+    rank_name: str
+    treedef_id: int
+
+    def to_json(self) -> Dict:
+        return {
+            "rank": self.rank_name,
+            "treedefId": self.treedef_id,
+        }
+
+    def to_key(self) -> Tuple[str, int]:
+        return (self.rank_name, self.treedef_id) # Check this line in unit test
+
+    def check_rank(self, tree: str) -> bool:
+        return TreeRank(self.rank_name, tree, self.treedef_id).check_rank()
+    
+    def validate_rank(self, tree) -> None:
+        TreeRank(self.rank_name, tree, self.treedef_id).validate_rank()
 
 class TreeRecord(NamedTuple):
     name: str
-    ranks: Dict[str, Dict[str, ColumnOptions]]
+    ranks: Dict[Union[str, TreeRankRecord], Dict[str, ColumnOptions]]
+    base_treedef_id: Optional[int] = None
 
     def apply_scoping(self, collection) -> "ScopedTreeRecord":
         from .scoping import apply_scoping_to_treerecord as apply_scoping
         return apply_scoping(self, collection)
 
     def get_cols(self) -> Set[str]:
-        return set(col.column for r in self.ranks.values() for col in r.values())
+        return {col.column for r in self.ranks.values() for col in r.values() if hasattr(col, 'column')}
 
     def to_json(self) -> Dict:
-        result = {
-            'ranks': {
-                rank: cols['name'].to_json() if len(cols) == 1 else dict(treeNodeCols={k: v.to_json() for k, v in cols.items()})
-                for rank, cols in self.ranks.items()
-            },
-        }
-        return { 'treeRecord': result }
+        # result: Dict[str, Union[str, int, Dict[str, Any]]] = {"ranks": {}}
+        result = {"ranks": {}} # type: ignore
+        
+        for rank, cols in self.ranks.items():
+            rank_key = rank.rank_name if isinstance(rank, TreeRankRecord) else rank # Check this line in unit test
+            treeNodeCols = {k: v.to_json() if hasattr(v, "to_json") else v for k, v in cols.items()}
+            
+            if len(cols) == 1:
+                result["ranks"][rank_key] = treeNodeCols["name"]
+            else:
+                rank_data = {"treeNodeCols": treeNodeCols}
+                if isinstance(rank, TreeRankRecord):
+                    rank_data["treeId"] = rank.treedef_id # type: ignore
+                result["ranks"][rank_key] = rank_data
+        
+        return {'treeRecord': result}
 
     def unparse(self) -> Dict:
         return { 'baseTableName': self.name, 'uploadble': self.to_json() }
 
 class ScopedTreeRecord(NamedTuple):
     name: str
-    ranks: Dict[str, Dict[str, ExtendedColumnOptions]]
+    ranks: Dict[TreeRankRecord, Dict[str, ExtendedColumnOptions]]
     treedef: Any
     treedefitems: List
     root: Optional[Any]
```

Comment:
simplify all this. what's the point of "TreeRankRecord"? Why is getting treedef and treedefitems based on basetreedef not enough? If we ever decide to support ranks across trees, you can message me "I told you so"

Reply:
> The front-end's design allowed for selecting ranks from multiple trees, so the back-end got adjusted to handle it.  It would have been better probably to just use the unique db id for Treedefid and Treedefitemid to identify ranks throughout Specify, rather than by name.

Reply:
The rank is identified by TreeRankRecord (rank_name, treedef_id) now instead of just rank_name.